l1-contracts/contracts/bridgehub/Bridgehub.sol

```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+import "./IBridgehub.sol";
+import "../bridge/interfaces/IL1Bridge.sol";
+import "../state-transition/IStateTransitionManager.sol";
+import "../common/ReentrancyGuard.sol";
+import "../state-transition/chain-interfaces/IZkSyncStateTransition.sol";
+import {ETH_TOKEN_ADDRESS} from "../common/Config.sol";
+
+contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2Step {
+    /// new fields
+    /// @notice we store registered stateTransitionManagers
+    mapping(address => bool) public stateTransitionManagerIsRegistered;
+    /// @notice we store registered tokens (for arbitrary base token)
+    mapping(address => bool) public tokenIsRegistered;
+    /// @notice we store registered bridges
+    mapping(address => bool) public tokenBridgeIsRegistered;
+
+    /// @notice chainID => stateTransitionManager contract address
+    mapping(uint256 => address) public stateTransitionManager;
+    /// @notice chainID => base token address
+    mapping(uint256 => address) public baseToken;
+    /// @notice chainID => bridge holding the base token
+    /// @notice a bridge can have multiple tokens
+    mapping(uint256 => address) public baseTokenBridge;
+
+    IL1Bridge public wethBridge;
+
+    modifier onlyBaseTokenBridge(uint256 _chainId) {
+        require(msg.sender == baseTokenBridge[_chainId], "Bridgehub: not base token bridge");
+        _;
+    }
+
+    function initialize(address _owner) external reentrancyGuardInitializer {
```

Comment:
This contract doesn't contain any delegate call, but the convention we have is to protect from the Parity hack is to add constructor as:

```
    constructor() reentrancyGuardInitializer { }
```